File: src/main/java/uk/gov/justice/digital/hmpps/oauth2server/resource/RedirectingLogoutSuccessHandler.java
package uk.gov.justice.digital.hmpps.oauth2server.resource;

import lombok.extern.slf4j.Slf4j;
import org.springframework.beans.factory.annotation.Value;
import org.springframework.security.core.Authentication;
import org.springframework.security.oauth2.provider.ClientDetailsService;
import org.springframework.security.web.authentication.logout.LogoutSuccessHandler;
import org.springframework.stereotype.Component;
import org.springframework.util.CollectionUtils;

import javax.servlet.http.HttpServletRequest;
import javax.servlet.http.HttpServletResponse;
import java.io.IOException;

@Component
@Slf4j
public class RedirectingLogoutSuccessHandler implements LogoutSuccessHandler {
    private final ClientDetailsService clientDetailsService;
    private final String servletContextPath;


    public RedirectingLogoutSuccessHandler(final ClientDetailsService clientDetailsService,
                                           @Value("#{servletContext.contextPath}") final String servletContextPath) {
        this.clientDetailsService = clientDetailsService;
        this.servletContextPath = servletContextPath;
    }

    @Override
    public void onLogoutSuccess(final HttpServletRequest request, final HttpServletResponse response, final Authentication authentication) throws IOException {
        final String client = request.getParameter("client_id");
        // If we have asked for a redirect, check it is valid for the client
        if (client != null) {
            final var clientDetails = clientDetailsService.loadClientByClientId(client);
            if (clientDetails != null && !CollectionUtils.isEmpty(clientDetails.getRegisteredRedirectUri())) {
                final var redirectUris = clientDetails.getRegisteredRedirectUri();
                final var redirect = request.getParameter("redirect_uri");
                if (redirectUris.contains(redirect)) {
                    response.sendRedirect(redirect);
                    return;
                }
                // second attempt - ignore or add trailing slash
                final var redirectSlash = redirect.endsWith("/") ? redirect.substring(0, redirect.length() - 1) : redirect + "/";
                if (redirectUris.contains(redirectSlash)) {
                    response.sendRedirect(redirectSlash);
                    return;
                }
                log.info("Ignoring redirect {} as not valid for client {}", redirect, client);
            }
        }
        response.sendRedirect(servletContextPath + "/login?logout");
    }
}

File: src/main/java/uk/gov/justice/digital/hmpps/oauth2server/resource/RedirectingLogoutSuccessHandler.kt
package uk.gov.justice.digital.hmpps.oauth2server.resource

import org.apache.commons.lang3.StringUtils
import org.slf4j.LoggerFactory
import org.springframework.beans.factory.annotation.Value
import org.springframework.security.core.Authentication
import org.springframework.security.oauth2.common.exceptions.RedirectMismatchException
import org.springframework.security.oauth2.provider.ClientDetails
import org.springframework.security.oauth2.provider.ClientDetailsService
import org.springframework.security.oauth2.provider.endpoint.DefaultRedirectResolver
import org.springframework.security.oauth2.provider.endpoint.RedirectResolver
import org.springframework.security.web.authentication.logout.LogoutSuccessHandler
import org.springframework.stereotype.Component
import org.springframework.util.CollectionUtils
import javax.servlet.http.HttpServletRequest
import javax.servlet.http.HttpServletResponse

@Suppress("DEPRECATION")
@Component

class RedirectingLogoutSuccessHandler(
  private val clientDetailsService: ClientDetailsService,
  @param:Value("#{servletContext.contextPath}") private val servletContextPath: String,
  @Value("\${application.authentication.match-subdomains}") matchSubdomains: Boolean
) : LogoutSuccessHandler {

  private val redirectResolver: RedirectResolver

  init {
    val defaultRedirectResolver = DefaultRedirectResolver()
    defaultRedirectResolver.setMatchSubdomains(matchSubdomains)
    redirectResolver = defaultRedirectResolver
  }

  companion object {
    private val log = LoggerFactory.getLogger(RedirectingLogoutSuccessHandler::class.java)
  }

  override fun onLogoutSuccess(
    request: HttpServletRequest,
    response: HttpServletResponse,
    authentication: Authentication?
  ) {
    val client = request.getParameter("client_id")
    val redirect = request.getParameter("redirect_uri")
    val error = request.getParameter("error")

    // If we have asked for a redirect, check it is valid for the client
    if (client != null && redirect != null) {
      val clientDetails = clientDetailsService.loadClientByClientId(client)
      if (clientDetails != null && !CollectionUtils.isEmpty(clientDetails.registeredRedirectUri)) {
        if (responseRedirectedOnValidRedirect(response, redirect, clientDetails)) {
          return
        }
        // second attempt - ignore or add trailing slash
        val redirectSlash = if (redirect.endsWith("/")) redirect.substring(0, redirect.length - 1) else "$redirect/"
        if (responseRedirectedOnValidRedirect(response, redirectSlash, clientDetails)) {
          return
        }
      }
    }
    response.sendRedirect(servletContextPath + "/login?logout" + if (StringUtils.isNotBlank(error)) "&error=$error" else "")
  }

  private fun responseRedirectedOnValidRedirect(
    response: HttpServletResponse,
    redirect: String,
    clientDetails: ClientDetails
  ): Boolean {
    try {
      response.sendRedirect(redirectResolver.resolveRedirect(redirect, clientDetails))
      return true
    } catch (rme: RedirectMismatchException) {
      log.info("Ignoring redirect {} as not valid for client {}, {}", redirect, clientDetails.clientId, rme.message)
    }
    return false
  }
}
